File: sdk/packages/shared/src/services/AutoSaveManager.ts
/**
 * White Room AutoSaveManager (TypeScript)
 *
 * Intelligent auto-save system with debouncing and periodic saves.
 * Prevents data loss while avoiding excessive writes.
 */

import { EventEmitter } from 'events';
import { AutoSaveRepository, AutoSave } from '../persistence/AutoSaveRepository';
import type { SongModel_v2 } from '../types/song-model';

// =============================================================================
// AUTO-SAVE MANAGER
// =============================================================================

export class AutoSaveManager extends EventEmitter {
  private autoSaveRepository: AutoSaveRepository;
  private debounceTimer: NodeJS.Timeout | null = null;
  private periodicTimer: NodeJS.Timeout | null = null;
  private pendingSave: {
    songId: string;
    song: SongModel_v2;
    timestamp: Date;
  } | null = null;

  public currentSongId: string | null = null;
  public lastSaveTime: Date | null = null;
  public isDirty: boolean = false;

  // Configuration
  private autoSaveEnabled = true;
  private debounceDelay = 2000;  // 2 seconds
  private periodicInterval = 60000;  // 1 minute
  private maxAutosaves = 10;

  constructor(autoSaveRepository: AutoSaveRepository) {
    super();
    this.autoSaveRepository = autoSaveRepository;
    this.startPeriodicTimer();
  }

  // ---------------------------------------------------------------------------
  // PUBLIC METHODS
  // ---------------------------------------------------------------------------

  /**
   * Mark song as dirty and schedule auto-save
   */
  public markDirty(song: SongModel_v2): void {
    this.isDirty = true;
    this.currentSongId = song.id;

    this.pendingSave = {
      songId: song.id,
      song: song,
      timestamp: new Date()
    };

    // Cancel existing timer
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
    }

    // Start new debounce timer
    this.debounceTimer = setTimeout(() => {
      if (this.autoSaveEnabled && this.pendingSave) {
        this.performAutoSave(this.pendingSave);
      }
    }, this.debounceDelay);
  }

  /**
   * Immediately save current song (skip debounce)
   */
  public saveNow(): void {
    if (!this.pendingSave) {
      throw new Error('No pending save available');
    }

    this.performAutoSave(this.pendingSave);
  }

  /**
   * Discard pending changes
   */
  public discardPendingSave(): void {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
      this.debounceTimer = null;
    }

    this.pendingSave = null;
    this.isDirty = false;
  }

  /**
   * Get all autosaves for current song
   */
  public getAutosaves(): AutoSave[] {
    if (!this.currentSongId) {
      throw new Error('No current song');
    }

    return this.autoSaveRepository.getAllForSong(this.currentSongId);
  }

  /**
   * Restore from autosave
   */
  public restoreFromAutosave(autosaveId: string): SongModel_v2 {
    const autosave = this.autoSaveRepository.read(autosaveId);

    if (!autosave) {
      throw new Error('Auto-save not found');
    }

    return JSON.parse(autosave.songJSON);
  }

  /**
   * Clear all autosaves for current song
   */
  public clearAutosaves(): void {
    if (!this.currentSongId) {
      throw new Error('No current song');
    }

    this.autoSaveRepository.deleteAllForSong(this.currentSongId);
  }

  // ---------------------------------------------------------------------------
  // PRIVATE METHODS
  // ---------------------------------------------------------------------------

  /**
   * Perform auto-save
   */
  private performAutoSave(pending: {
    songId: string;
    song: SongModel_v2;
    timestamp: Date;
  }): void {
    try {
      const autosave: AutoSave = {
        id: this.generateUUID(),
        songId: pending.songId,
        songJSON: JSON.stringify(pending.song),
        timestamp: pending.timestamp,
        description: this.generateDescription(pending.song)
      };

      this.autoSaveRepository.create(autosave);

      this.lastSaveTime = new Date();
      this.isDirty = false;
      this.pendingSave = null;

      this.pruneOldAutosaves(pending.songId);

      this.emit('autosave', autosave);

      console.log(`Auto-saved song: ${pending.song.metadata.title}`);
    } catch (error) {
      this.emit('error', error);
      console.error('Auto-save failed:', error);
    }
  }

  /**
   * Start periodic timer
   */
  private startPeriodicTimer(): void {
    this.periodicTimer = setInterval(() => {
      if (this.autoSaveEnabled && this.isDirty && this.pendingSave) {
        this.performAutoSave(this.pendingSave);
      }
    }, this.periodicInterval);
  }

  /**
   * Prune old autosaves
   */
  private pruneOldAutosaves(songId: string): void {
    const autosaves = this.autoSaveRepository.getAllForSong(songId);

    if (autosaves.length > this.maxAutosaves) {
      // Sort by timestamp, oldest first
      const sorted = autosaves.sort((a, b) =>
        a.timestamp.getTime() - b.timestamp.getTime()
      );

      const toDelete = sorted.slice(this.maxAutosaves);

      for (const autosave of toDelete) {
        this.autoSaveRepository.delete(autosave.id);
      }

      console.log(`AutoSaveManager: Pruned ${toDelete.length} old autosaves`);
    }
  }

  /**
   * Generate description for autosave
   */
  private generateDescription(song: SongModel_v2): string {
    const timestamp = new Date().toLocaleString();
    const trackCount = song.mixGraph.tracks.length;
    const sectionCount = song.sections.length;

    return `Auto-save - ${timestamp} - ${trackCount} tracks, ${sectionCount} sections`;
  }

  /**
   * Generate UUID
   */
  private generateUUID(): string {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
      const r = Math.random() * 16 | 0;
      const v = c === 'x' ? r : (r & 0x3 | 0x8);
      return v.toString(16);
    });
  }

  // ---------------------------------------------------------------------------
  // CLEANUP
  // ---------------------------------------------------------------------------

  public destroy(): void {
    if (this.debounceTimer) {
      clearTimeout(this.debounceTimer);
      this.debounceTimer = null;
    }

    if (this.periodicTimer) {
      clearInterval(this.periodicTimer);
      this.periodicTimer = null;
    }

    this.removeAllListeners();
  }
}

// =============================================================================
// ERROR TYPES
// =============================================================================

export class AutoSaveError extends Error {
  constructor(message: string, public code: string) {
    super(message);
    this.name = 'AutoSaveError';
  }
}

File: swift_frontend/SwiftFrontendShared/Services/AutoSaveManager.swift
/**
 * White Room AutoSaveManager
 *
 * Intelligent auto-save system with debouncing and periodic saves.
 * Prevents data loss while avoiding excessive writes.
 *
 * Features:
 * - Debounced saves (2-second delay after changes)
 * - Periodic saves (every 60 seconds if dirty)
 * - Automatic pruning (max 10 autosaves per song)
 * - Restore from any autosave
 * - Thread-safe with actor
 */

import Foundation

// MARK: - AutoSave Manager

/// Manages auto-save operations with debouncing and periodic saves
public actor AutoSaveManager: ObservableObject {
  // MARK: - Properties

  private let autoSaveRepository: AutoSaveRepository
  private let songRepository: SongRepository
  private var debounceTimer: Task<Void, Never>?
  private var periodicTimer: Task<Void, Never>?
  private var pendingSave: PendingSave?

  public var currentSongId: String?
  @Published public var lastSaveTime: Date?
  @Published public var isDirty: Bool = false

  private struct PendingSave {
    let songId: String
    let song: Song
    let timestamp: Date
  }

  // MARK: - Configuration

  private var autoSaveEnabled: Bool { true }
  private var debounceDelay: TimeInterval { 2.0 }  // 2 seconds
  private var periodicInterval: TimeInterval { 60.0 }  // 1 minute
  private var maxAutosaves: Int { 10 }

  // MARK: - Initialization

  public init(
    autoSaveRepository: AutoSaveRepository,
    songRepository: SongRepository
  ) {
    self.autoSaveRepository = autoSaveRepository
    self.songRepository = songRepository
    startPeriodicTimer()
  }

  // MARK: - Public Methods

  /// Mark song as dirty and schedule auto-save
  public func markDirty(_ song: Song) async {
    isDirty = true
    currentSongId = song.id

    let pending = PendingSave(
      songId: song.id,
      song: song,
      timestamp: Date()
    )

    // Store pending save
    self.pendingSave = pending

    // Cancel existing timer
    debounceTimer?.cancel()

    // Start new debounce timer
    debounceTimer = Task {
      do {
        try await Task.sleep(for: .seconds(debounceDelay))

        if !Task.isCancelled {
          try? await performAutoSave(pending)
        }
      } catch {
        // Timer cancelled or error
      }
    }
  }

  /// Immediately save current song (skip debounce)
  public func saveNow() async throws {
    guard let pending = pendingSave else {
      throw AutoSaveError.noPendingSave
    }

    try await performAutoSave(pending)
  }

  /// Discard pending changes
  public func discardPendingSave() {
    debounceTimer?.cancel()
    pendingSave = nil
    isDirty = false
  }

  /// Get all autosaves for current song
  public func getAutosaves() async throws -> [AutoSave] {
    guard let songId = currentSongId else {
      throw AutoSaveError.noCurrentSong
    }

    return try await autoSaveRepository.getAllForSong(songId)
  }

  /// Restore from autosave
  public func restoreFromAutosave(_ autosaveId: String) async throws -> Song {
    guard let autosave = try await autoSaveRepository.read(id: autosaveId) else {
      throw AutoSaveError.autosaveNotFound
    }

    // Decode song from JSON
    let songData = autosave.songJSON.data(using: .utf8)!
    return try JSONDecoder().decode(Song.self, from: songData)
  }

  /// Clear all autosaves for current song
  public func clearAutosaves() async throws {
    guard let songId = currentSongId else {
      throw AutoSaveError.noCurrentSong
    }

    try await autoSaveRepository.deleteAllForSong(songId)
  }

  // MARK: - Private Methods

  private func performAutoSave(_ pending: PendingSave) async throws {
    // Encode song to JSON
    let songData = try JSONEncoder().encode(pending.song)
    let songJSON = String(data: songData, encoding: .utf8)!

    // Create autosave
    let autosave = AutoSave(
      id: UUID().uuidString,
      songId: pending.songId,
      songJSON: songJSON,
      timestamp: pending.timestamp,
      description: generateDescription(pending.song)
    )

    // Save to database
    try await autoSaveRepository.create(autosave)

    // Update state
    lastSaveTime = Date()
    isDirty = false
    self.pendingSave = nil

    // Prune old autosaves
    try await pruneOldAutosaves(for: pending.songId)

    NSLog("Auto-saved song: \(pending.song.name)")
  }

  private func startPeriodicTimer() {
    periodicTimer = Task {
      while !Task.isCancelled {
        do {
          try await Task.sleep(for: .seconds(periodicInterval))

          if autoSaveEnabled, isDirty, let pending = pendingSave {
            try? await performAutoSave(pending)
          }
        } catch {
          // Timer cancelled or error
          break
        }
      }
    }
  }

  private func pruneOldAutosaves(for songId: String) async throws {
    let autosaves = try await autoSaveRepository.getAllForSong(songId)

    if autosaves.count > maxAutosaves {
      // Sort by timestamp, oldest first
      let sorted = autosaves.sorted { $0.timestamp < $1.timestamp }
      let toDelete = sorted.dropFirst(maxAutosaves)

      for autosave in toDelete {
        try await autoSaveRepository.delete(id: autosave.id)
      }

      NSLog("AutoSaveManager: Pruned \(toDelete.count) old autosaves")
    }
  }

  private func generateDescription(_ song: Song) -> String {
    let formatter = DateFormatter()
    formatter.dateStyle = .short
    formatter.timeStyle = .medium

    let timestamp = formatter.string(from: Date())
    let trackCount = song.trackConfigs.count
    let sectionCount = song.sections.count

    return "Auto-save - \(timestamp) - \(trackCount) tracks, \(sectionCount) sections"
  }

  // MARK: - Cleanup

  deinit {
    debounceTimer?.cancel()
    periodicTimer?.cancel()
  }
}

// MARK: - AutoSave Error

public enum AutoSaveError: LocalizedError {
  case noPendingSave
  case noCurrentSong
  case autosaveNotFound

  public var errorDescription: String? {
    switch self {
    case .noPendingSave:
      return "No pending save available"
    case .noCurrentSong:
      return "No current song to save"
    case .autosaveNotFound:
      return "Auto-save not found"
    }
  }
}
